library/core/src/hint.rs

```diff
@@ -382,6 +391,85 @@ pub fn spin_loop() {
 ///
 /// This makes our benchmark much more realistic to how the function would actually be used, where
 /// arguments are usually not known at compile time and the result is used in some way.
+///
+/// # How to use this
+///
+/// In practice, `black_box` serves two purposes:
+///
+/// 1. It prevents the compiler from making optimizations related to the value of the returned
+///    type
+/// 2. It forces the input to be calculated, even if its results are never used
+///
+/// ```
+/// use std::hint::black_box;
+///
+/// let zero = 0;
+/// let five = 5;
+///
+/// // The compiler will see this and remove the `* five` call, because it knows that multiplying
+/// // any integer by 0 will result in 0. This is a value optimization: the compiler knows the
+/// // value of `zero` must be 0, and thus can make optimizations related to that.
```

Comment:
```suggestion
/// // any integer by 0 will result in 0.
```
"value optimization" is not a term we use in the compiler and I've never seen it before. This optimization specifically is called constant propagation. You could call it that, specifically "This is constant propagation" instead of "This is a value optimization", but I don't think this second sentence adds much, the first is quite clear on its own.